edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/authenticators/DeviceScopeCertificateAuthenticator.cs

```diff
@@ -0,0 +1,234 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Hub.CloudProxy.Authenticators
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Net;
+    using System.Security.Cryptography;
+    using System.Security.Cryptography.X509Certificates;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Common.Data;
+    using Microsoft.Azure.Devices.Common.Security;
+    using Microsoft.Azure.Devices.Edge.Hub.Core;
+    using Microsoft.Azure.Devices.Edge.Hub.Core.Device;
+    using Microsoft.Azure.Devices.Edge.Hub.Core.Identity;
+    using Microsoft.Azure.Devices.Edge.Hub.Core.Identity.Service;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Logging;
+
+    public class DeviceScopeCertificateAuthenticator : IAuthenticator
+    {
+        readonly IDeviceScopeIdentitiesCache deviceScopeIdentitiesCache;
+        readonly string iothubHostName;
+        readonly IAuthenticator underlyingAuthenticator;
+        readonly IList<X509Certificate2> trustBundle;
+
+        public DeviceScopeCertificateAuthenticator(
+            IDeviceScopeIdentitiesCache deviceScopeIdentitiesCache,
+            string iothubHostName,
+            IAuthenticator underlyingAuthenticator,
+            IList<X509Certificate2> trustBundle)
+        {
+            this.deviceScopeIdentitiesCache = Preconditions.CheckNotNull(deviceScopeIdentitiesCache, nameof(deviceScopeIdentitiesCache));
+            this.underlyingAuthenticator = Preconditions.CheckNotNull(underlyingAuthenticator, nameof(underlyingAuthenticator));
+            this.iothubHostName = Preconditions.CheckNonWhiteSpace(iothubHostName, nameof(iothubHostName));
+            this.trustBundle = Preconditions.CheckNotNull(trustBundle, nameof(trustBundle));
+        }
+
+        public Task<bool> AuthenticateAsync(IClientCredentials clientCredentials)
+            => this.AuthenticateAsync(clientCredentials, false);
+
+        public Task<bool> ReauthenticateAsync(IClientCredentials clientCredentials)
+            => this.AuthenticateAsync(clientCredentials, true);
+
+        async Task<bool> AuthenticateAsync(IClientCredentials clientCredentials, bool reAuthenticating)
+        {
+            if (!(clientCredentials is ICertificateCredentials certificateCredentials))
+            {
+                return false;
+            }
+
+            Option<ServiceIdentity> serviceIdentity = await this.deviceScopeIdentitiesCache.GetServiceIdentity(clientCredentials.Identity.Id, reAuthenticating);
+            if (serviceIdentity.HasValue)
+            {
+                // currently authenticating modules via X.509 is disabled. all the necessary pieces to authenticate
+                // modules via X.509 CA are implemented below and to enable modules to authenticate remove this check
+                if (certificateCredentials.Identity is IModuleIdentity)
+                {
```

Comment:
Is there a reason we are doing this, other than to reduce the surface area for testing?

Reply:
In short yes. Resolving this for now per our convo.